### ISSUE TITLE
[FEATURE] MS Teams Compatible Messages

### ISSUE THREAD

**Issue by apsaras** [OpenSIPS/opensips#1647]
Hello

I am working for few days now to implementing a Microsoft Teams Compatible SBC based on OpenSIPs. Teams is the replacement for Skype for Business which enables the use of a Plugin called Direct Routing, in order to interconnect your Teams Infrastructure with PSTN Providers.

Teams SBC and signaling requirements are not well documented, hence I am working on trial and error to find the correct syntax of the messages to be sent.

From my reading and research up to now, it seams that Teams SBC is using Contact Header to identify the sender. For example OPTIONS should have a Contact Header with FQDN like <sip:teams.mycompany.com:5060;transport=tls>, TO and FROM header should contain the FQDN as well and all should match the TLS certification.

The same applies to INVITE towards Teams.

My feature request is the following

We should add Contact Header on OPTIONS which will include the IP or FQDN as defined on listening interface or as defined by an AVP.

On INVITE sent to Teams, I am using topology hiding which generates a Contact Header containing the IP address of the listening interface. That should be changed to FQDN as defined on listening interface or defined by an AVP.

To summarize for now (until my next finding)
- Add Contact Header to OPTIONS with FQDN
- Use FQDN on Topology Hiding Contact Headers

Best regards

**Comment by vasilevalex** [comment]
Hi @apsaras But all, what you wrote is very simple to implement without any changes. I have working MS Teams integration.
1) Add Contact Header to OPTIONS:
```
local_route {
	$var(src) = "pstnhub.microsoft.com";
	if (is_method("OPTIONS") && ($(ru{s.index, $var(src)}) != NULL)) {
		$var(contact) = "sip:FQDN";
		append_hf("Contact: <$var(contact):5061;transport=tls>\r\n");
	}
}
```
2) INVITE towards MS Teams doesn't require FQDN in Contact. (I have working integration, don't believe Oracle and AudioCodes)

TLS is the challenge, but it was resolved yesterday with commit: 44394e322eb2c52565629c6889d15e9e35837b39

**Comment by apsaras** [comment]
Hello @vasilevalex .

That is true. That is the way I implemented my self but my point is to enable modules to do so without local manipulation.

For me INVITE towards MS is not working with result 403 and error "<IP> is not a valid FQDN" so my guess was that contact is the issue.

Is the script you have shareable?

**Comment by vasilevalex** [comment]
RFC3261 Section 11.1 https://tools.ietf.org/html/rfc3261#section-11.1
"A Contact header field MAY be present in an OPTIONS."
So there is absolutely no reason to hardcode it in module. If you want - just add it on script level.
"IP is not a valid FQDN" - you should change RR.

**Comment by apsaras** [comment]
Thank you @vasilevalex for your help.

Can you please share few more info?

Are you using dispatcher for routing because I am not able to route using TLS

ERROR:core:tcp_connect_blocking_timeout: timeout 99151 ms elapsed from 100000 s
ERROR:proto_tls:tls_sync_connect: tcp_blocking_connect failed
ERROR:proto_tls:proto_tls_send: connect failed

**Comment by netaskd** [comment]
try to use `set_advertised_address("you.fqdn.domain");` in route script before send request.
or set it globally via `advertised_address="your.fqdn.domain"`

**Comment by bogdan-iancu** [comment]
Hi guys - is this discussion concluded ? is there anything left to be discussed / done on the OpenSIPS side ?

**Comment by apsaras** [comment]
Hello Bogdan.

As @vasilevalex mentioning, most of the MS Teams requirement can probably be applied using routing logic. Unfortunately that didn't work for me, but that's another discussion.

You can close that.

**Comment by vasilevalex** [comment]
@apsaras Finally, there are some notes about Teams: https://blog.opensips.org/2019/09/16/opensips-as-ms-teams-sbc/

**Comment by billkim** [comment]
@vasilevalex My opensips proto_tls failed to send error below, I wonder what the problem is...
I've configured below.
-----------------------------------------------------------------
tcp_connect_timeout=3000
modparam("tls_mgm", "tls_handshake_timeout", 3000)
modparam("proto_tls", "tls_max_msg_chunks", 8)
-----------------------------------------------------------------

opensips in debug mode displays these logs..

OPTIONS sip:sip.pstnhub.microsoft.com SIP/2.0
Via: SIP/2.0/TLS 112.221.--.---:5061;branch=z9hG4bK21b8.75ea7966.0
To: sip:sip.pstnhub.microsoft.com
From: <sip:probe@sbc.lime------.--->;tag=121a9e246cd95656de30c2d389c7dd83-7d2b
CSeq: 14 OPTIONS
Call-ID: 3583b4a07ae01cf1-34825@172.16.104.101
Max-Forwards: 70
Content-Length: 0
User-Agent: OpenSIPS (2.4.6 (x86_64/linux))
Contact: <sip:sbc.lime------.---:5061;transport=tls>


Nov 11 05:52:12 [34825] ERROR:proto_tls:proto_tls_send: failed to send
Nov 11 05:52:12 [34825] DBG:core:tcpconn_release:  releasing con 0x7f88784eddd0, state -2, fd=-1, id=276844070
Nov 11 05:52:12 [34825] DBG:core:tcpconn_release:  extra_data 0x7f8878506fa8
Nov 11 05:52:12 [34825] ERROR:tm:msg_send: send() to 52.114.7.24:5061 for proto tls/3 failed
Nov 11 05:52:12 [34825] ERROR:tm:t_uac: attempt to send to 'sip:sip.pstnhub.microsoft.com' failed
Nov 11 05:52:12 [34825] DBG:tm:insert_timer_unsafe: [0]: 0x7f88784c7028 (55)
Nov 11 05:52:12 [34828] DBG:core:handle_tcp_worker: response= 7f88784eddd0, -2 from tcp worker 34825 (0)
Nov 11 05:52:12 [34825] DBG:core:handle_io: We have received conn 0x7f88784eddd0 with rw 1 on fd 5
Nov 11 05:52:12 [34825] DBG:core:io_watch_add: [TCP_worker] io_watch_add op (5 on 31) (0x55ad1c328ea0, 5, 19, 0x7f88784eddd0,1), fd_no=4/6553
Nov 11 05:52:12 [34828] DBG:core:tcpconn_destroy: delaying (0x7f88784eddd0, flags 0018) ref = 1 ...
Nov 11 05:52:12 [34825] DBG:core:io_watch_del: [TCP_worker] io_watch_del op on index -1 5 (0x55ad1c328ea0, 5, -1, 0x10,0x1) fd_no=5 called
Nov 11 05:52:12 [34825] DBG:core:tcpconn_release:  releasing con 0x7f88784eddd0, state -3, fd=-1, id=276844070
Nov 11 05:52:12 [34825] DBG:core:tcpconn_release:  extra_data 0x7f8878506fa8
Nov 11 05:52:12 [34828] DBG:core:handle_tcp_worker: response= 7f88784eddd0, -3 from tcp worker 34825 (0)
Nov 11 05:52:12 [34828] DBG:core:tcpconn_destroy: destroying connection 0x7f88784eddd0, flags 0018
Nov 11 05:52:12 [34828] DBG:proto_tls:tls_conn_clean: entered
Nov 11 05:52:12 [34828] DBG:proto_tls:tls_update_fd: New fd is 38

**Comment by vasilevalex** [comment]
Hi @billkim Have you read the article? There are all settings, that you must change. In your log I can see that connection failed, but I can't see the reason.

**Comment by apsaras** [comment]
Hello team.

Based on the new multi-tenant configuration for MS Teams, it is required to change Contact Header to much customers' domain.

I am using topology hiding but I can not find a way to dynamically specify the FQDN to be used.

Any suggestion?

https://docs.microsoft.com/en-us/microsoftteams/direct-routing-sbc-multiple-tenants#deploy-and-configure-the-sbc

"Subdomains MUST match the FQDN name of the trunk that will be configured for the customer and the FQDN in the Contact header when sending the Invite to Office 365.

When a call arrives at the Office 365 Direct Routing interface, the interface uses the Contact header to find the tenant where the user should be looked up. Direct Routing does not use phone number lookup on the Invite, as some customers might have non-DID numbers that can overlap in several tenants. Therefore, the FQDN name in the Contact header is required to identify the exact tenant to look up the user by the phone number."

**Comment by apsaras** [comment]
Resolved using set_advertised_address

**Comment by billkim** [comment]
@apsaras thank you for sharing your exprience.

**Comment by arsperger** [comment]
Hi guys,

I ran into the same problem as @apsaras  in a multi-domain scenario. 
Maybe someone can shed a bit light on how is possible to pass FQDN  dynamically into local_route so that Opensips send OPTIONS from different domains in the Contact header?
I use dispatcher module to OPTIONS pinging (keep ms-teams sip addresses only) and drouting for routing.

**Comment by arsperger** [comment]
> Hi guys,
> 
> I ran into the same problem as @apsaras in a multi-domain scenario.
> Maybe someone can shed a bit light on how is possible to pass FQDN dynamically into local_route so that Opensips send OPTIONS from different domains in the Contact header?
> I use dispatcher module to OPTIONS pinging (keep ms-teams sip addresses only) and drouting for routing.

I've sorted it out with correct setting of advertised_address.

**Comment by celya44** [comment]
Hello,

I tried to make teams direct routing working but I don't anderstand why it doesn' work.

The TLS seem working 
INFO:proto_tls:tls_connect: New TLS connection to 52.114.76.76:5061 established
INFO:proto_tls:tls_dump_cert_info: tls_connect: server TLS certificate subject: /CN=sip.pstnhub.microsoft.com, issuer: /C=US/ST=Washington/L=Redmond/O=Microsoft Corporation/OU=Microsoft IT/CN=Microsoft IT TLS CA 4
INFO:proto_tls:tls_dump_cert_info: tls_connect: local TLS client certificate subject: /CN=opensips-XXXXXXX.XXXX.fr, issuer: /C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3

My OPTION seem good
OPTIONS sip:sip.pstnhub.microsoft.com SIP/2.0
Via: SIP/2.0/TLS XXX.XXX.XXX.XXX:5061;branch=z9hG4bKb89a.926d26a.0
To: sip:sip.pstnhub.microsoft.com
From: <sip:XXXX-teams.XXXXX.fr:5061;transport=tls>;tag=d450e4f35227cb2b93484a73ea135e23-15cf
CSeq: 14 OPTIONS
Call-ID: 2cd5a0bc6227e066-18908@XXX.XXX.XXX.XXX
Max-Forwards: 70
Content-Length: 0
User-Agent: OpenSIPS (3.0.2 (x86_64/linux))
Contact: <sip:XXX-teams.XXXXX.fr:5061;transport=tls>

But I never receiv 200 OK or some thing else and I fall in failure_route.
Any idea ?

**Comment by apsaras** [comment]
Hello

does XXX-teams.XXXXX.fr resolves back to your server?

Is 5061 TCP opened at your firewall?

**Comment by celya44** [comment]
Yes, I check the name of the direct routing sbc and the FQDN in the contact, it is the same and I can resolve it.    
I checked my FW and Any it open from all the MS network and I have no reject frame.

**Comment by celya44** [comment]
I solved my problem, I generated a new certificate for the FQDN, and now I get answers from MS.

**Comment by dzhou121** [comment]
@celya44 I've got the same issue with no response from Teams at all. Can I ask which certificate were you using and which new certificate fixed the issue?

**Comment by celya44** [comment]
It's letsencrypt certificate. But I also alow in the teams console my domaine.
